flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java

```diff
@@ -636,6 +637,15 @@ public ResourceProfile getFreeResourceOf(InstanceID instanceID) {
         return taskManagerTracker.getRegisteredResourceOf(instanceID);
     }
 
+    @Override
+    public Collection<SlotInfo> getAllocatedSlotsOf(InstanceID instanceID) {
+        return taskManagerTracker.getRegisteredTaskManager(instanceID)
+                .map(TaskManagerInfo::getAllocatedSlots).map(Map::values)
+                .orElse(Collections.emptyList()).stream()
+                .map(slot -> new SlotInfo(slot.getJobId(), slot.getResourceProfile()))
+                .collect(Collectors.toList());
```

Comment:
This will scale terribly. Do we really need information for every single slot?
Why not slots + number of slots of that type? Or total acquired resources? Or available/total resources of the TM?

Reply:
Currently, we already have the available/total resources of the TM. For fine-grained resource management, slots are dynamically created with different resource profiles. It would be good to let user know the what type of slot is allocated and which job it is allocated for. I don't think it will help allocate to reorganize it to slots + number of slots of that type. In the worst case, that will increase the data size instead.

After an offline discussion with @vthinkxie , I think we can add only this information to `TaskManagerDetailsInfo`.

Reply:
But why bother with slots in the first place? Isn't the whole point for fine-grained resource management to not really bother with slots but _resources_ instead, so why then expose it as slots again in the UI?

Why not just say "Job A has allocated X memory and Y CPUs on TM Z to run N tasks."

What value does it bring to know this instead:
"Job as has N slots on TM Z, where slot 1 uses X_1 memory any Y_1 cpus, and slot 2 uses X_2 memory and Y_2 cpus, and ..."

Reply:
"Job as has N slots on TM Z, where slot 1 uses X_1 memory any Y_1 cpus, and slot 2 uses X_2 memory and Y_2 cpus, and ..." could be aggregated and get "Job A has allocated X memory and Y CPUs on TM Z to run N tasks."
while it cannot be pushed backward.
Since TM and resources are implemented on a slots basis, the interface layer should expose atomic properties

Reply:
I think how to expose that information to the user is depends on the WebUI and might be out of the scope of this PR. In current interface, we give the UI flexibility to choose how to aggregate it.
Also, user might want to know the explicit resource of each task, especially for external resources.